File: daily/25(Reverse Nodes In K Groups)/Solution.py
class Solution:
    def reverseKGroup(self, head, k):
        if not head or k < 1:
            return head
        count = 0
        current = head
        while current:
            count += 1
            current = current.next
        if count < k:
            return head
        prev = None
        current = head
        for _ in range(k):
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        head.next = self.reverseKGroup(current, k)
        return prev
k=Solution()
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
# Creating a linked list for testing
def create_linked_list(values):
    head = ListNode(values[0])
    current = head
    for value in values[1:]:
        current.next = ListNode(value)
        current = current.next
    return head
def print_linked_list(head):
    current = head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")
# Test the reverseKGroup function
head = create_linked_list([1,2,3,4,5])
k_value = 3
reversed_head = k.reverseKGroup(head, k_value)
print_linked_list(reversed_head)  

    

File: daily/25(Reverse Nodes In K Groups)/Solution.java
class Solution {
    class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
        ListNode(int x, ListNode next) {
            val = x;
            this.next = next;
        }
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k <= 1) return head;

        ListNode dummy = new ListNode(0, head);
        ListNode prevGroup = dummy;
        ListNode tail = head;

        int index = 0;

        while (tail != null) {
            index++;
            if (index % k == 0) {
                ListNode start = prevGroup.next;
                ListNode nextGroup = tail.next;

                // Reverse k nodes between start and tail (inclusive)
                reverse(start, tail);

                prevGroup.next = tail;
                start.next = nextGroup;

                prevGroup = start;
                tail = nextGroup;
            } else {
                tail = tail.next;
            }
        }

        return dummy.next;
    }

    private void reverse(ListNode start, ListNode end) {
        ListNode prev = null;
        ListNode curr = start;
        ListNode stop = end.next;

        while (curr != stop) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
    }
}